modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSCoverageReader.java

```diff
@@ -133,6 +133,32 @@ public WMSCoverageReader(WebMapServer wms, Layer layer, String style) {
         }
     }
 
+    public WMSCoverageReader(WebMapServer wms, Layer layer, String style, String preferredFormat) {
+        this.wms = wms;
+
+        // init the reader
+        addLayer(layer, style);
+
+        // best guess at the format with a preference for PNG (since it's normally transparent)
+        List<String> formats = wms.getCapabilities().getRequest().getGetMap().getFormats();
+        this.format = preferredFormat;
+        // verify if prefferedFormat is supported
+        if (!formats.contains(preferredFormat)) {
+            // if prefereded format is not supported default to first available
+            LOGGER.warning(preferredFormat + " format is not supported, falling back to defaults");
+            for (String format : formats) {
```

Comment:
This code is an exact copy paste of the code just 20 lines above, can we avoid this :fearful: ?

Reply:
sure I will move this code into a utility function